src/libre/ast_new_from_fsm.c

```diff
@@ -0,0 +1,474 @@
+	free(edge);
+}
+
+static void
+free_restate(struct rese *state)
+{
+	assert(state);
+	assert(state->from == state && state->to == state);
+
+	while (state->succnext != state)
+		free_reedge(state->succnext);
+	while (state->prednext != state)
+		free_reedge(state->prednext);
+
+	ast_expr_free(state->expr);
+	state->expr = NULL;
+}
+
+static void
+free_restates(struct rese *states, unsigned int numstates)
+{
+	unsigned int i;
+
+	for (i = 0; i < numstates; i++)
+		free_restate(&states[i]);
+
+	free(states);
+}
+
+static int
+remove_state(struct rese *state)
+{
+	struct rese *p /*pred*/, *ps /*predsucc*/, *s /*succ*/;
+
+	assert(state);
+	assert(state->from == state->to);
+
+	/* build new edges for all pred->succ combinations
+	 *
+	 * if pred->self is A, self->self is B, self->succ is C, we add
+	 * AB*C to the pred->succ transition */
+	for (p = state->prednext; p != state; p = p->prednext) {
+		assert(p->from != state && p->to == state);
+
+		ps = p->from->succnext;
+
+		for (s = state->succnext; s != state; s = s->succnext) {
+			struct rese *e;
+			struct ast_expr *cat;
+			struct ast_expr *tmp;
+
+			assert(s->from == state && s->to != state);
+
+			if (p->from == s->to) {
+				e = p->from;
+			} else {
+				while (ps->from != ps->to && ps->to < s->to)
+					ps = ps->succnext;
+
+				if (ps->to != s->to)
+					ps = build_reedge(p->from, s->to);
+
+				e = ps;
+			}
+
+			assert(p->expr);
+			assert(s->expr);
+
+			cat = ast_make_expr_concat(0);
+			if (!cat) {
+				return 0;
+			}
+
+			tmp = p->expr;
+			if (!ast_expr_clone(&tmp)) {
+				ast_expr_free(cat);
+				return 0;
+			}
+
+			if (!ast_add_expr_concat(cat, tmp)) {
+				ast_expr_free(tmp);
+				ast_expr_free(cat);
+				return 0;
+			}
+
+			if (state->expr) {
+				struct ast_expr *rep;
+				tmp = state->expr;
+				if (!ast_expr_clone(&tmp)) {
+					ast_expr_free(cat);
+					return 0;
+				}
+				rep = ast_make_expr_repeat(0, tmp,
+					ast_make_count(0, NULL, AST_COUNT_UNBOUNDED, NULL));
+				if (!rep) {
+					ast_expr_free(tmp);
+					ast_expr_free(cat);
+					return 0;
+				}
+				if (!ast_add_expr_concat(cat, rep)) {
+					ast_expr_free(rep);
+					ast_expr_free(cat);
+					return 0;
+				}
+			}
+
+			tmp = s->expr;
+			if (!ast_expr_clone(&tmp)) {
+				ast_expr_free(cat);
+				return 0;
+			}
+			if (!ast_add_expr_concat(cat, tmp)) {
+				ast_expr_free(tmp);
+				ast_expr_free(cat);
+				return 0;
+			}
+
+			if (!e->expr) {
+				e->expr = cat;
+			} else {
+				if (!change_to_alt(&e->expr)) {
+					ast_expr_free(cat);
+					return 0;
+				}
+
+				if (!ast_add_expr_alt(e->expr, cat)) {
+					ast_expr_free(cat);
+					return 0;
+				}
+			}
+		}
+	}
+
+	free_restate(state);
+}
+
+struct ast *
+ast_new_from_fsm(const struct fsm *fsm)
+{
+	struct ast *ast;
+	struct ast_expr *expr;
+	struct fsm_options opt;
+
+	unsigned int numstates;
+	fsm_state_t start;
+	struct rese *restates;
+	struct rese *restart, *reend;
+
+	unsigned int i;
+	struct rese *state;
+
```

Comment:
Could you assert() on the contractual constraints here please? and comment them in the header.
Or maybe there are none, I'm not sure!

Reply:
It should work for arbitrary FSMs, even though apparently it does not yet :) I added an assert that `fsm` must not be a null pointer, I think you typically leave that one out of any header comments, but I can add it if you like.

Reply:
Yes, no need to comment that :)

I think we must assume that there's a reachable end state perhaps?

Reply:
Oh no, we'd accept that and produce a tombstone. nice